feat: FIT-974: [FSM API] Manual Transition Endpoint

matt-bernstein opened the pull request:
<!-- Follow Merge -->
This PR was automatically generated via Follow Merge. [Workflow Run](https://github.com/HumanSignal/label-studio/actions/runs/19518304411)
Please ensure that all linked upstream Pull Requests are merged before proceeding with this one.
- https://github.com/HumanSignal/label-studio-sdk/pull/676
<!-- Follow Merge -->

Add api to transition entities between states
<!--

This description MUST be filled out for a PR to receive a review. Its primary purposes are:

 - to enable your reviewer to review your code easily, and
 - to convince your reviewer that your code works as intended.

Some pointers to think about when filling out your PR description:
 - Reason for change: Description of problem and solution
 - Screenshots: All visible changes should include screenshots.
 - Rollout strategy: How will this code be rolled out? Feature flags / env var / other
 - Testing: Description of how this is being verified
 - Risks: Are there any known risks associated with this change, eg to security or performance?
 - Reviewer notes: Any info to help reviewers approve the PR
 - General notes: Any info to help onlookers understand the code, or callouts to significant portions.

You may use AI tools such as Copilot Actions to assist with writing your PR description (see https://docs.github.com/en/copilot/using-github-copilot/using-github-copilot-for-pull-requests/creating-a-pull-request-summary-with-github-copilot); however, an AI summary isn't enough by itself. You'll need to provide your reviewer with strong evidence that your code works as intended, which requires actually running the code and showing that it works.

-->

matt-bernstein committed:
feat: FIT-974: [FSM API] Manual Transition Endpoint

matt-bernstein committed:
update urls

matt-bernstein committed:
update tests

matt-bernstein committed:
change redundant test

matt-bernstein committed:
move test to lse

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-FIT-974

matt-bernstein committed:
bugfix

matt-bernstein committed:
Revert "bugfix"

This reverts commit 1b969eab1126dc07141059bb7b1c2823449ae514.

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19515568700

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-FIT-974'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19515568700

matt-bernstein committed:
make test LSO only

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-FIT-974

matt-bernstein commented:
/fmt
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19516764422)

matt-bernstein committed:
Apply pre-commit linters

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19516764422

mcanu commented on label_studio/fsm/api.py:
```diff
@@ -99,3 +106,117 @@ def list(self, request, *args, **kwargs):
+            raise ValidationError({'detail': str(e)})
+        except Exception as e:
+            # Unexpected failure
+            raise APIException(detail='Transition execution failed') from e
```
I wouldn't catch these, these look like 5XX errors that we'll return to the user and catch them in Sentry. The detail of the exception isn't showed to the user

mcanu commented on label_studio/fsm/api.py:
```diff
@@ -99,3 +106,117 @@ def list(self, request, *args, **kwargs):
+    def post(self, request, *args, **kwargs):
+        entity_name = kwargs['entity_name']
+        if entity_name not in state_model_registry.get_all_models():
+            raise NotFound()
```
Can we extract this to a mixin that is shared by the history endpoint as well?

mcanu commented on label_studio/fsm/api.py:
```diff
@@ -99,3 +106,117 @@ def list(self, request, *args, **kwargs):
+            raise ValidationError({'detail': str(e)})
+        except TransitionValidationError as e:
+            # Explicit validation failure
+            raise ValidationError({'detail': str(e), **(e.context or {})})
```
q: What do this look like? Just a warning that we don't show unwanted information to the user

matt-bernstein replied:
Here's the only time TransitionValidationError is used with context outside of a test:
https://github.com/HumanSignal/label-studio/blob/9c9bf2024783bfe1560d27d3006f53a7ba8c9804/label_studio/fsm/transitions.py#L291-L297

It's definitely possible that this could leak unwanted info in the future, but as of now, it's very rare to use `context` in the first place, and when it's used it's for obvious information. Here's an example from a test:
```
raise TransitionValidationError('Invalid assignee ID', {'assignee_id': self.assignee_id})
```

I'm torn on this one, it could make debugging much easier to have, but there is no guardrail to stop us from writing something sensitive there even though it seems unlikely

mcanu replied:
I would not include the context, for now at least. We can log it in a warning log there to help us for troubleshooting. But I agree we don't know what can be in the context, and there might be some internal stuff we don't want to show.

mcanu replied:
I'd defer to any public info worth showing to the user for the Exception message

mcanu commented on label_studio/fsm/api.py:
```diff
@@ -99,3 +106,117 @@ def list(self, request, *args, **kwargs):
+                'new_state': state_record.state,
+                'state_record': state_record_data,
+            }
+        return Response(response_payload, status=status.HTTP_200_OK)
```
Lets use the response serializer her

matt-bernstein committed:
review comments

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19518304411

matt-bernstein committed:
fix refactor

matt-bernstein committed:
don't swallow context in common error handling layer

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19538594387)
> Successfully merged:  4 files changed, 20 insertions(+)

matt-bernstein committed:
Merge branch 'develop' into 'fb-FIT-974'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19538594387

mcanu commented on label_studio/fsm/tests/test_transition_api.py:
I would move this to `fsm/tests/test_api.py`

bmartel reviewed (approved):
Looks great, thanks @matt-bernstein 🙌

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-FIT-974

matt-bernstein committed:
move tests

matt-bernstein committed:
logging error context

matt-bernstein committed:
lint

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio-enterprise/actions/runs/19549626609
